### PR TITLE
Reading IR models from buffer

### LINKED ISSUE #15578: Read ir model from buffer not implemented
- OpenCV => 4.1.1
- Operating System / Platform => Windows 64 Bit / Ubuntu 18.04 LTS
- Compiler => Visual Studio 2017 / clang6.0

We are trying to load an IR model from a memory buffer (loaded from a single binary file, not relevant) but `dnn::readNet()` for dldt from buffer is not implemented.

```cpp
 else if (framework == "dldt")
        CV_Error(Error::StsNotImplemented, "Reading Intel's Model Optimizer models from buffers");
```

My colleague is already working on submitting a pull request to implement this.

### PR DESCRIPTION
**Pull request by Quantizs**
resolves #15578

### This pullrequest changes

Added implementations to load IR models from buffer

### CODE REVIEW

**Comment by dkurt** [comment]
@Quantizs, hi!

Please change a base branch to 3.4 and choose a different name for feature branch:

```
git checkout 3.4
git checkout -b branch_name
```

File: modules/dnn/src/dnn.cpp [hunk 1]
```diff
@@ -2875,6 +2875,87 @@ Net Net::readFromModelOptimizer(const String& xml, const String& bin)
...
+        ld.backendNodes[DNN_BACKEND_INFERENCE_ENGINE] = backendNode;
+
+        for (int i = 0; i < inputsNames.size(); ++i)
+            cvNet.connect(0, i, lid, i);
+    }
+    cvNet.setPreferableBackend(DNN_BACKEND_INFERENCE_ENGINE);
+
+    cvNet.impl->skipInfEngineInit = true;
+    return cvNet;
+}
```
Comment (dkurt) [inline modules/dnn/src/dnn.cpp]: Please create a separate method which returns `cv::Net` from `reader`. Let's reduce code duplication.

**Comment by marcellBan** [comment]
@dkurt hi,

We are working on the requested changes, but as the merged branch cannot be changed (just the base) should we open a new PR? (We kind of missed the branch naming guidelines during our first read trough.)

**Comment by dkurt** [comment]
@marcellBan, it's more about general recommendation, not OpenCV only. If you add a commit to a local `master` branch, then you just cannot keep this branch aligned with a remote one. So it's better to create a separate branches for local changes but keep origin branches unchanged for robust `pull` procedures.

**Comment by marcellBan** [comment]
@dkurt, I totally understand and agree with it as I've worked with git for a few years now but I was asking if we should open a new PR with a different branch or should we just use this one and solve any upcoming git issues on our end.

**Comment by dkurt** [comment]
@marcellBan, If it's OK to modify the branch names `master` locally, please try this:

1) rename `master` to something temporal like `fix`
2) go to branch `3.4`
3) `git checkout -b master`
4) `git cherry-pick your_commit_hash` or just copy-paste the code and create a new commit
5) push resulting branch with `-f` to the fork
6) reopen this PR. Click `Edit` button where you can change target branch to `3.4`

**Comment by Quantizs** [comment]
We rebased the forked branch on 3.4 and I have refactored the code as requested and pushed it to `Quantizs:master` but i can't reopen the pull request or edit it.

**Comment by dkurt** [comment]
@Quantizs, can you please try to open a new one?

**Comment by Quantizs** [comment]
@dkurt Reopen under #16034

**Pull request closed without merging**